docs/tutorial/tests/steps/library_steps.py

```diff
@@ -1,34 +1,72 @@
+    step_data_table_titles = [cell.value for cell in data_table.rows[0].cells]
+    assert step_data_table_titles == ["Author", "Title"]
 
+    author_and_title_list = [[cell.value for cell in row.cells] for row in data_table.rows[1:]]
 
-@given("these books in the catalog")
-def these_books_in_the_catalog(step):
-    context.catalog = Catalog()
-    context.catalog.add_books_to_catalog(step.data_table)
+    books = [Book(author, title) for author, title in author_and_title_list]
+    return books
 
 
-@when(parsers.re("a (?P<search_type>name|title) search is performed for " +
-"(?P<search_term>.+)"))
-def a_SEARCH_TYPE_is_performed_for_SEARCH_TERM(search_type, search_term):
+# Steps to be used in scenarios are defined with special decorators
+@given(
+    "these books in the catalog",
+    # Steps are allowed to inject new fixtures or overwrite existing ones
+    target_fixture="catalog",
+)
+def these_books_in_the_catalog(
+    # `step` fixture is injected by pytest DI mechanism into scope of step by default
+    step: Step,
+):
+    catalog = Catalog()
+
+    books = get_books_from_data_table(step.data_table)
+    catalog.add_books_to_catalog(books)
+    yield catalog
+
+
+@when(
+    # Step definitions could have parameters. Here could be raw stings, cucumber expressions or regular expressions
+    re.compile('a (?P<search_type>name|title) search is performed for "(?P<search_term>.+)"'),
+    target_fixture="search_results",
+)
+def a_search_type_is_performed_for_search_term(
+    # `search_results` is a usual pytest fixture defined somewhere else and injected by pytest DI mechanism.
+    # In this case it will be provided by conftest.py
+    search_results,
+    # `search_type` and `search_term` are parameters of this step and are injected by step definition
+    search_type: Literal["name", "title"],
+    search_term: str,
+    # `catalog` is a fixture injected by another step
+    catalog: Catalog,
+):
     if search_type == "title":
-        raise NotImplementedError("Title searches are not yet implemented.")
-    context.search_results = context.catalog.search_by_author(search_term)
+        search_results.extend(catalog.search_by_title(search_term))
+    elif search_type == "name":
+        search_results.extend(catalog.search_by_author(search_term))
+    else:
+        assert False, "Unknown"
+
+    yield search_results
 
 
 @then("only these books will be returned")
-def only_these_books_will_be_returned(step):
-expected_books = context.catalog.read_books_from_table(step.data_table)
-verify_returned_books(context.search_results, expected_books)
+def only_these_books_will_be_returned(
+    # fixtures persist during step execution, so usual context is not required,
+    # so if you define fixture dependencies debugging becomes much easier.
+    search_results,
+    step: Step,
+    catalog: Catalog,
+):
+    expected_books = get_books_from_data_table(step.data_table)
+    assert all([book in catalog.storage for book in expected_books])
```

Comment:
This appears to verify that every book in 'expected_books' is also in 'catalog.storage'; I have two questions:
1) Why are we verifying against catalog.storage rather than verifying against search_results?
2) Does this also verify that everything that is in search_results is also in expected_books? That is, if we expect one book, and that book is indeed in catalog.storage; this test will pass. However, what if two books were returned - one expected book and one unexpected book? Will this test catch that case?